src/crimson/common/interruptible_future.h

```diff
@@ -0,0 +1,729 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <seastar/core/future-util.hh>
+#include <seastar/core/do_with.hh>
+#include <seastar/core/when_all.hh>
+
+#include "crimson/common/errorator.h"
+
+namespace crimson::interruptible {
+
+struct ready_future_marker {};
+struct exception_future_marker {};
+
+template <typename InterruptCond>
+class interruptible_future_builder;
+
+template <typename InterruptCond>
+struct interruptor;
+
+template <typename InterruptCond>
+using InterruptCondRef = seastar::lw_shared_ptr<InterruptCond>;
+
+template <typename InterruptCond>
+InterruptCondRef<InterruptCond> interrupt_cond;
+
+template <typename InterruptCond, typename FutureType>
+class interruptible_future_detail {};
+
+template <typename FutureType>
+struct is_interruptible_future : public std::false_type {};
+
+template <typename InterruptCond, typename FutureType>
+struct is_interruptible_future<
+  interruptible_future_detail<
+    InterruptCond,
+    FutureType>>
+  : public std::true_type {};
+
+template <typename InterruptCond, typename Func, typename... T>
+auto call_with_interruption(
+  InterruptCondRef<InterruptCond> interrupt_condition,
+  Func&& func, T&&... args)
+{
+  using Result = std::invoke_result_t<Func, T...>;
+  using futurator_t = seastar::futurize<Result>;
+  if constexpr (sizeof...(T) == 1) {
+    if constexpr (InterruptCond::template is_interruption_v<
+	  std::tuple_element_t<0, std::tuple<T...>>>) {
+      return futurator_t::make_exception_future(
+	  std::get<0>(std::tuple(std::forward<T>(args)...)));
+    }
+    if constexpr (seastar::is_future<
+	  std::tuple_element_t<0, std::tuple<T...>>>::value) {
+      auto fut = std::get<0>(std::tuple(std::forward<T>(args)...));
+      std::exception_ptr eptr = fut.get_exception();
+      if (fut.failed() && interrupt_condition->is_interruption(eptr)) {
+	return futurator_t::make_exception_future(std::move(eptr));
+      }
+    }
+  }
+
+  if (interrupt_condition) {
+    auto [interrupt, fut] = interrupt_condition->template may_interrupt<
+      typename futurator_t::type>();
+    if (interrupt) {
+      return std::move(*fut);
+    }
+    interrupt_cond<InterruptCond> = interrupt_condition;
+  }
+
+  if constexpr (std::is_same_v<Result, void>) {
+    func(std::forward<T>(args)...);
```

Comment:
this branch fails to return a value. please note, a `future<>` is returned at line 52 and line 60.